partiql-eval/src/eval.rs

```diff
@@ -164,6 +227,21 @@ impl Evaluable for EvalJoin {
             })
         }
 
+        #[inline]
+        fn cross_lateral<'a>(
+            left: impl Iterator<Item = &'a Value> + Clone + 'a,
+            right: impl Iterator<Item = &'a Value> + Clone + 'a,
+        ) -> impl Iterator<Item = Tuple> + 'a {
+            left.zip(right).map(|(l_tuple, r_tuple)| {
```

Comment:
Interesting that we can `.zip` the two iterators together to perform the lateral cross join rather than `.cartesian_product` for the non-lateral cross join. Guess it makes sense based on my current understanding. I'll be looking more into lateral joins this week to see if this implementation would need altered.

Reply:
Looking at this now with some other test cases, I don't think `.zip` is the right behavior here. It might have worked for the `subquery_in_from` example you had in `lib.rs` since the `lhs` and `rhs` have the same length of elements (i.e. they both use `data`). But if the `lhs` and `rhs` have different lengths, then some rows may be ignored.

Consider a test like the following taken from [spec example 10](https://partiql.org/assets/PartiQL-Specification.pdf#subsection.5.3):
```Rust
#[test]
fn select_with_cross_join_sensors() {
    let mut lg = LogicalPlan::new();

    let from_lhs = lg.add_operator(scan("sensors", "s"));
    let from_rhs = lg.add_operator(BindingsExpr::Scan(logical::Scan {
        expr: path_var("s", "readings"),
        as_key: "r".to_string(),
        at_key: None,
    }));

    let project = lg.add_operator(Project(logical::Project {
        exprs: HashMap::from([
            ("v".to_string(), path_var("r", "v")),
        ]),
    }));

    let join = lg.add_operator(BindingsExpr::Join(logical::Join {
        kind: JoinKind::CrossLateral,
        on: None,
    }));

    let sink = lg.add_operator(BindingsExpr::Sink);
    lg.add_flow_with_branch_num(from_lhs, from_rhs, 0);
    lg.add_flow_with_branch_num(from_lhs, join, 0);
    lg.add_flow_with_branch_num(from_rhs, join, 1);
    lg.add_flow_with_branch_num(join, project, 0);
    lg.add_flow_with_branch_num(project, sink, 0);

    let out = evaluate(lg, join_data_sensors());
    println!("{:?}", &out);
}
```
Using `join_data_sensors()` specified as follows:
```Rust
fn join_data_sensors() -> MapBindings<Value> {
    let sensors = partiql_list![
        partiql_tuple![
            ("readings", partiql_list![
                partiql_tuple![("v", 1.3)],
                partiql_tuple![("v", 2)],
            ])
        ],
        partiql_tuple![
            ("readings", partiql_list![
                partiql_tuple![("v", 0.7)],
                partiql_tuple![("v", 0.8)],
                partiql_tuple![("v", 0.9)],
            ])
        ],
    ];
    let mut bindings = MapBindings::default();
    bindings.insert("sensors", sensors.into());
    bindings
}
```

The above outputs:
```
<<[Tuple { v: 1.3 }, Tuple { v: 2 }]>>
```

Rather than the `v` for every `reading` (i.e.

```
<<[Tuple { v: 1.3 }, Tuple { v: 2 }, Tuple { v: 0.7 }, Tuple { v: 0.8 }, Tuple { v: 0.9 }]>>
```

Reply:
Nice clarification! honestly I had my doubts about the current `CrossLateral` that's why I mentioned in the PR description that it needs a revisit. The main objective of this PR is to implement subqueries but not joins. I know as a follow-on we're already looking into `LATERAL JOIN` implementation in general.